### ISSUE TITLE
Next epoch ledger returns unconfirmed ledger

### ISSUE THREAD

**Issue by garethtdavies** [MinaProtocol/mina#8246]
It seems https://github.com/MinaProtocol/mina/pull/7612 no longer works perhaps after refactoring to the ledger export commands. Was this an intentional change?

Example: At block height 2873 at global slot 7405 in epoch 1 the command `mina ledger export `next-epoch-ledger` returns a staking ledger. However, the last block of epoch 0 was 2762 and so the ledger has not been finalized and I would expect it to return a message that it was not yet available?

**Comment by psteckler** [comment]
I've verified that, in `master`, the error checking code introduced in #7612 is still run by `export next-epoch-ledger`.

**Comment by garethtdavies** [comment]
I should have said this was on `finalfinal2` which I've now removed. I will try to reproduce on `devnet` at start of the new epoch.

**Comment by garethtdavies** [comment]
Devnet just went into epoch 2 and yet `mina ledger export next-epoch-ledger` returns a staking ledger. I was expecting it to return an error message?

**Comment by garethtdavies** [comment]
After some further testing, it is returning the current staking ledger. Here was the status of the node:

```
Global number of accounts:                     1670
Block height:                                  4616
Max observed block height:                     4616
Max observed unvalidated block height:         4616
Local uptime:                                  20h32m21s
Ledger Merkle root:                            jxL4KLYgYCqiuo862fawrDvUTyXNv3FN3pGVV3R5275UjGoA1zL
Protocol state hash:                           3NLpRTcn9LpUWekbsjtevZmKWrGhTNdovesKV5LayoWLBnNYkGaJ
Chain id:                                      8af43cf261ea10c761ec540f92aafb76aec56d8d74f77c836f3ab1de5ce4eac5
Git SHA-1:                                     68200c7b409e105d174f079c5c39f7926819784b
Configuration directory:                       /root/.mina-config
Peers:                                         28
User_commands sent:                            0
SNARK worker:                                  None
SNARK work fee:                                100000000
Sync status:                                   Synced
Catchup status:                                
        To build breadcrumb:           0
        To initial validate:           0
        Finished:                      94
        To verify:                     0
        To download:                   0
        Waiting for parent to finish:  0

Block producers running:                       0
Best tip consensus time:                       epoch=1, slot=59
Best tip global slot (across all hard-forks):  7199
```

Export the next staking ledger (expect this to return not available):

`mina ledger export next-epoch-ledger > test.json`

This returns a staking ledger. So getting the hash of this:

```
mina ledger hash --ledger-file test.json
jxmLNQcQ5Tgbf6z7828m4RL5SDsW7XTFfQK6QsQ9quanJSzTJcN
```

From GraphQL we can see that this is the current staking ledger. 

```
"blocks": [
      {
        "protocolState": {
          "consensusState": {
            "epoch": 1,
            "nextEpochData": {
              "ledger": {
                "hash": "jxkx5tG7if37WCYybvfP1aP27Ai9MDz6LLMjwnzFEbZPTLiECH5"
              }
            },
            "slot": 61,
            "stakingEpochData": {
              "ledger": {
                "hash": "jxmLNQcQ5Tgbf6z7828m4RL5SDsW7XTFfQK6QsQ9quanJSzTJcN"
              }
            }
          }
        }
      }
    ]
  }
```

**Comment by garethtdavies** [comment]
Running this at block 4864, so once the next ledger is finalized does correctly return the ledger hash `jxkx5tG7if37WCYybvfP1aP27Ai9MDz6LLMjwnzFEbZPTLiECH5`

**Comment by psteckler** [comment]
Here is the calculation to determine whether the next epoch ledger is finalized: https://github.com/MinaProtocol/mina/blob/master-qa/src/lib/mina_lib/mina_lib.ml#L795

According to this calculation, it's finalized if the root and best tip are in the same epoch, or if the root is in the genesis epoch.

**Comment by garethtdavies** [comment]
> According to this calculation, it's finalized if the root and best tip are in the same epoch, or if the root is in the genesis epoch.

So could that be the issue as it thought the next ledger was finalized as it was referring to the genesis ledger hash? So this won't happen for future epochs and it will correctly state that the next ledger hash isn't yet available (if it is not finalized)

**Comment by psteckler** [comment]
> So could that be the issue as it thought the next ledger was finalized as it was referring to the genesis ledger hash? So this won't happen for future epochs and it will correctly state that the next ledger hash isn't yet available (if it is not finalized)

If I understand, when you ran the export in epoch 1, the finalized root was in the genesis epoch, falling into the special case. (I think that's a bug, given my limited understanding.) That shouldn't happen again, now that the net is beyond the genesis epoch.

**Comment by garethtdavies** [comment]
Yes, exactly. Obviously, this will be a consideration for the upcoming mainnet epoch 1 transition. Might also be a factor post any hard-fork where there is a genesis block. Given the ledger isn't finalized when the "wrong" output is returned it's more a potential point of confusion for this first epoch.

**Comment by psteckler** [comment]
@deepthiskumar Can you check whether Gareth's explanation makes sense?

**Comment by deepthiskumar** [comment]
> @deepthiskumar Can you check whether Gareth's explanation makes sense?

Yeah, this should be- finalized if the root and best tip are in the same epoch, or if the **best tip** is in the genesis epoch. I'll push a fix.

**Comment by garethtdavies** [comment]
This was completed
